app/gui/src/dashboard/layouts/AssetPanel/components/AssetVersions.tsx

```diff
@@ -65,81 +76,88 @@ function AssetVersionsInternal(props: AssetVersionsInternalProps) {
+
+    if (options?.start === true && newAsset != null && item.type === AssetType.project) {
+      // This is SAFE because we know that the the new asset is a Project,
+      // because we can't create a duplicate with a different type.
+      // eslint-disable-next-line no-restricted-syntax
+      await openProjectLocally(newAsset as ProjectAsset, backend.type)
+    }
+  })
+
+  const doRestore = useEventCallback((version: Version) =>
+    restoreMutation.mutateAsync({
+      versionId: version.versionId,
+      placeholderId: S3ObjectVersionId(uniqueString()),
+    }),
+  )
+
+  if (versions.length === 0) {
+    return <Result status="info" centered title={getText('noVersionsFound')} />
+  }
+
+  if (latestVersion == null) {
+    return <Result status="error" centered title={getText('fetchLatestVersionError')} />
+  }
+
   return (
-    <div className="pointer-events-auto flex flex-1 shrink-0 flex-col items-center overflow-y-auto overflow-x-hidden">
-      {versionsQuery.data.length === 0 ?
-        <div>{getText('noVersionsFound')}</div>
-      : latestVersion == null ?
-        <Text color="danger">{getText('fetchLatestVersionError')}</Text>
-      : [
-          ...placeholderVersions.map((version, i) => (
-            <AssetVersion
-              key={version.versionId}
-              placeholder
-              number={versionsQuery.data.length + placeholderVersions.length - i}
-              version={version}
-              item={item}
-              backend={backend}
-              latestVersion={latestVersion}
-              doRestore={noop}
-            />
-          )),
-          ...versionsQuery.data.map((version, i) => (
-            <AssetVersion
-              key={version.versionId}
-              number={versionsQuery.data.length - i}
-              version={version}
-              item={item}
-              backend={backend}
-              latestVersion={latestVersion}
-              doRestore={() =>
-                restoreMutation.mutateAsync({
-                  versionId: version.versionId,
-                  placeholderId: S3ObjectVersionId(uniqueString()),
-                })
-              }
-            />
-          )),
-        ]
-      }
+    <div className="flex w-full flex-col">
+      {versions.map((version, index) => (
+        <div key={version.versionId}>
+          <AssetVersion
+            version={version}
+            otherVersions={versions}
+            item={item}
+            backend={backend}
+            previousVersion={versions[index + 1]}
+            doRestore={doRestore}
+            doDuplicate={doDuplicate}
+          />
+
+          {index !== versions.length - 1 && <div className="ml-[3px] h-5 w-[0.5px] bg-primary" />}
+        </div>
+      ))}
     </div>
   )
 }
+
+/**
+ * Check if the asset is allowed to have versions.
+ */
+function isAllowedAssetType(asset: AnyAsset): asset is DatalinkAsset | FileAsset | ProjectAsset {
+  return ALLOWED_ASSET_TYPES.includes(asset.type)
```

Comment:
optional: consider making `ALLOWED_ASSET_TYPES` an `as const` and then using `includes()` or `includesPredicate()` from `#/utilities/array`

Reply:
I just inlined the `ALLOWED_ASSET_TYPES`